lib/FrontendTool/FrontendTool.cpp

```diff
@@ -237,12 +237,154 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
 }
 }
 
+namespace {
+
+// The most interesting/tricky thing here is _which_ paths get recorded. It
+// depends on how the module was synthesized. The key points are:
+//
+// 1. Paths in the cache are not recorded - Precondition: the corresponding
+//    path to the swiftinterface must already be present as a key in
+//    pathToModuleDecl.
+// 2. swiftmodules next to a swiftinterface are saved if they are up-to-date.
+void computeSwiftModuleTraceInfo(
+    const SmallPtrSetImpl<ModuleDecl *> &importedModules,
+    const llvm::DenseMap<StringRef, ModuleDecl *> &pathToModuleDecl,
+    const DependencyTracker &depTracker,
+    StringRef prebuiltCachePath,
+    std::vector<SwiftModuleTraceInfo> &traceInfo) {
+
+  SmallString<256> buffer;
+
+  std::string errMsg;
+  llvm::raw_string_ostream err(errMsg);
+
+  auto errorUnexpectedPath = [&err, &pathToModuleDecl]() {
+    err << "The module <-> path mapping we have is:\n";
+    for (auto &m: pathToModuleDecl)
+      err << m.second->getName() << " <-> " << m.first << '\n';
+    llvm::report_fatal_error(err.str());
+  };
+
+  using namespace llvm::sys;
+
+  auto computeAdjacentInterfacePath = [](SmallVectorImpl<char> &modPath) {
+    auto swiftInterfaceExt =
+      file_types::getExtension(file_types::TY_SwiftModuleInterfaceFile);
+    path::replace_extension(modPath, swiftInterfaceExt);
+  };
+
+  for (auto &depPath : depTracker.getDependencies()) {
+
+    // Decide if this is a swiftmodule based on the extension of the raw
+    // dependency path, as the true file may have a different one.
+    // For example, this might happen when the canonicalized path points to
+    // a Content Addressed Storage (CAS) location.
+    auto moduleFileType =
+      file_types::lookupTypeForExtension(path::extension(depPath));
+    auto isSwiftmodule =
+      moduleFileType == file_types::TY_SwiftModuleFile;
+    auto isSwiftinterface =
+      moduleFileType == file_types::TY_SwiftModuleInterfaceFile;
+
+    if (isSwiftmodule || isSwiftinterface) {
+      auto dep = pathToModuleDecl.find(depPath);
+      if (dep != pathToModuleDecl.end()) {
```

Comment:
Can you use early exits more? Both this and the containing `if` seem like they can get the error handling out of the way and `continue`.

Reply:
I can get rid of the outer one but this one seems necessary, because in both cases, we have some complicated logic. For the outer-one, there was nothing to be done, if the condition was false, so we can simply `continue`.

Reply:
I see what you're saying. Instead of 

```
if (good) {
  happy()
  continue;
}
handleError();
```

I should change it to

```
if (!good) {
  handleError();
  continue;
}

happy();
```

Reply:
Either way is fine if they're both complicated, I gues. I thought the error path was going to be shorter here but maybe I'm misreading it.